libyul/YulName.cpp

```diff
@@ -0,0 +1,112 @@
+{
+	auto const emptyName = defineName("");
+	yulAssert(emptyName == YulNameRepository::emptyName());
+}
+
+std::string_view YulNameRepository::requiredLabelOf(YulName const _name) const
+{
+	auto const label = labelOf(_name);
+	yulAssert(label.has_value(), "YulName currently has no defined label in the YulNameRepository.");
+	return label.value();
+}
+
+std::optional<YulNameRepository::YulName> YulNameRepository::nameOfLabel(std::string_view const _label) const
+{
+	auto const it = std::find(m_definedLabels.begin(), m_definedLabels.end(), _label);
+	if (it != m_definedLabels.end())
+	{
+		YulName labelName{static_cast<YulName>(std::distance(m_definedLabels.begin(), it))};
+		// mostly it'll be iota
+		if (!isDerivedName(labelName) && std::get<0>(m_names[static_cast<size_t>(labelName)]) == labelName)
+			return std::get<0>(m_names[static_cast<size_t>(labelName)]);
+		// if not iota, we have to search
+		auto itName = std::find(m_names.rbegin(), m_names.rend(), std::make_tuple(labelName, YulNameState::DEFINED));
+		if (itName != m_names.rend())
+			return std::get<0>(*itName);
+	}
+	return std::nullopt;
+}
+
+void YulNameRepository::generateLabels(std::set<YulName> const& _usedNames, std::set<std::string> const& _illegal)
+{
+	std::set<std::string> used;
+	std::set<YulName> toDerive;
+	for (auto const name: _usedNames)
+	{
+		if (!isDerivedName(name))
+		{
+			auto const label = labelOf(name);
+			yulAssert(label.has_value());
+			auto const [it, emplaced] = used.emplace(*label);
+			if (!emplaced || _illegal.count(*it) > 0)
+				// there's been a clash ,e.g., by calling generate labels twice;
+				// let's remove this name and derive it instead
+				toDerive.insert(name);
+		}
+	}
+
+	std::vector<std::tuple<std::string, YulName>> generated;
+	auto namesIt = _usedNames.begin();
+	for (size_t nameValue = 1; nameValue < m_names.size(); ++nameValue)
+	{
+		auto name = static_cast<YulName>(nameValue);
+		if (namesIt != _usedNames.end() && name == *namesIt)
+		{
+			if (isDerivedName(name) || toDerive.find(name) != toDerive.end())
+			{
+				std::string const baseLabel(baseLabelOf(name));
+				std::string label (baseLabel);
+				size_t bump = 1;
+				while (used.count(label) > 0 || _illegal.count(label) > 0)
+					label = fmt::format(FMT_COMPILE("{}_{}"), baseLabel, bump++);
+				if (auto const existingDefinedName = nameOfLabel(label); existingDefinedName.has_value())
+				{
+					std::get<0>(m_names[static_cast<size_t>(name)]) = *existingDefinedName;
+					std::get<1>(m_names[static_cast<size_t>(name)]) = YulNameState::DEFINED;
+				}
+				else
+					generated.emplace_back(label, name);
+				used.insert(label);
+			}
+			++namesIt;
+		}
+	}
+
+	for (auto const& [label, name] : generated)
+	{
+		yulAssert(_illegal.count(label) == 0);
+		m_definedLabels.emplace_back(label);
+		std::get<0>(m_names[static_cast<size_t>(name)]) = static_cast<YulName>(m_definedLabels.size() - 1);
+		std::get<1>(m_names[static_cast<size_t>(name)]) = YulNameState::DEFINED;
```

Comment:
This is repeated in `defineName()` as well. Would be better to define a helper:

```suggestion
		convertToDefined(name, label);
```